### PR TITLE
fix(promises): add default handler for all unhandled promises

### PR DESCRIPTION
**Pull request by trieloff**
@tripodsan in https://github.com/adobe/helix-dispatch/pull/353#issuecomment-711113626
> something changed and now uncaught promises exit the process. this is a problem, since in the fetcher race, we ignore the promises as soon as 1 successful. also see #326

I assume the CircleCI container got upgraded to [node 15](https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V15.md#throw-on-unhandled-rejections---33021)

This is a brute force fix that restores the old behavior

### CODE REVIEW

**Comment by trieloff** [comment]
That's not the issue and not the fix

**Comment by trieloff** [comment]
`git bisect` points here:

https://github.com/adobe/helix-dispatch/pull/328/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556R127

**Comment by trieloff** [comment]
Mocha now enforces unhandled promises, so I've made sure they get handled.

**Comment by tripodsan** [comment]
> Mocha now enforces unhandled promises, so I've made sure they get handled.

yes. it's mocha 8.2, we see this in other project.

File: src/index.js [hunk 1]
```diff
@@ -213,11 +207,24 @@ async function executeActions(params) {
...
-    log.error(`error while invoking fetchers: ${e}`);
+    log.error(`!!!error while invoking fetchers: ${e}`);
     /* istanbul ignore next */
     return {
       error: `${String(e.stack)}`,
     };
+  } finally {
+    try {
+      // we need to wait for the dummy 404 request, otherwise we have unhandled promise rejections
+      await rejection;
```
Comment (tripodsan) [inline src/index.js:218]: why is this needed?

File: src/index.js [hunk 2]
```diff
@@ -213,11 +207,24 @@ async function executeActions(params) {
     // a fetchers `resolve` should never throw an exception but report a proper status response.
     // so we consider any exception thrown as application error and propagate it to openwhisk.
     /* istanbul ignore next */
-    log.error(`error while invoking fetchers: ${e}`);
+    log.error(`!!!error while invoking fetchers: ${e}`);
```
Comment (tripodsan) [inline src/index.js:210]: ?

Comment (trieloff) [inline src/index.js:210]:
```suggestion
    log.error(`error while invoking fetchers: ${e}`);
```

File: src/index.js (hunk 1 above)
Comment (trieloff) [reply src/index.js:218]: `fetch404Promise` which we resolve below gets initialized with `rejected`, but might get re-assigned, in which case `rejected` is a dangling promise.

Comment (tripodsan) [reply src/index.js:218]: as I wrote in helix-chat on slack, I don't think that this is enough. we have to catch the exceptions when they occur, and not eventually.....

File: .circleci/config.yml [hunk 3]
```diff
@@ -2,7 +2,7 @@ version: 2.1
 executors:
   node10:
     docker:
-      - image: circleci/node:10
+      - image: circleci/node:15
```
Comment (tripodsan) [inline .circleci/config.yml]: I don't think this is a good idea :-)

**Review by tripodsan — CHANGES_REQUESTED** [review]
I disagree with the fix and the change in circle ci to use node:15

File: src/index.js (hunk 1 above)
Comment (trieloff) [reply src/index.js:218]: Really not sure how to move forward here. The change so far seems to fix the issue at hand, but I sense you have a bigger refactor in mind.

File: .circleci/config.yml (hunk 3 above)
Comment (trieloff) [reply .circleci/config.yml]: Already reverted, I just wanted to see what would happen.

File: src/index.js (hunk 1 above)
Comment (tripodsan) [reply src/index.js:218]: I did some testing, and I think the following works:

```const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

async function test() {
  console.log('test sleeping...');
  await sleep(100);
  throw Error('overslept.');
}
async function run() {
  const task = test();
  task.catch(() => {
    console.log('error'); // catch the rejection when it's generated
  });
  console.log('run sleeping...');
  await sleep(200);
  const result = await task; // this will throw another rejection.
  return result;
}
run().catch(console.error);
```

try adding the `.catch(() => {})` to the `promise404`

File: src/index.js [hunk 4]
```diff
@@ -138,7 +138,7 @@ async function executeActions(params) {
       });
   });
 
-  let fetch404Promise = Promise.reject();
+  let fetch404Promise = Promise.reject().catch(() => {});
```
Comment (tripodsan) [inline src/index.js:141]: maybe also below?
https://github.com/adobe/helix-dispatch/pull/355/files#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556L152

Comment (trieloff) [reply src/index.js:141]: Missing a line number

**Review by tripodsan — APPROVED** [review]
